File: frontend/lib/utils/deriveUsername.ts
export function deriveUsername(
  displayName: string,
  existingUsernames?: Set<string>
): string {
  let base = displayName
    .trim()
    .toLowerCase()
    .replace(/\s+/g, '_')
    .replace(/[^\w]/g, '')
    .replace(/_+/g, '_')
    .replace(/^_|_$/g, '');

  if (!existingUsernames) return base;

  let username = base;
  let counter = 1;
  while (existingUsernames.has(username)) {
    username = `${base}_${counter}`;
    counter++;
  }
  return username;
}

File: mobile/lib/core/utils/derive_username.dart
/// Mirrors `frontend/lib/utils/deriveUsername.ts`.
String deriveUsername(String displayName, {Set<String>? existingUsernames}) {
  var base = displayName
      .trim()
      .toLowerCase()
      .replaceAll(RegExp(r'\s+'), '_')
      .replaceAll(RegExp(r'[^\w]'), '')
      .replaceAll(RegExp(r'_+'), '_')
      .replaceAll(RegExp(r'^_|_$'), '');

  if (existingUsernames == null) return base;

  var username = base;
  var counter = 1;
  while (existingUsernames.contains(username)) {
    username = '${base}_$counter';
    counter++;
  }
  return username;
}